File: 43.R
# Sub-string divisibility

get_divisible_by <- function(divisor) {
    numbers <- vector()
    number <- 0
    while ( number + divisor < 1000 ) {
        number <- number + divisor
        str_number <- sprintf("%03d", number)
        num_uniq_digits <- length(unique(strsplit(str_number, split=NULL)[[1]]))
        if ( num_uniq_digits == nchar(str_number) ) {
            numbers <- c(numbers, str_number)
        }
    }

    return(numbers)
}

filter_not_pandigital <- function(candidates) {
    filtered <- vector()
    for ( candidate in candidates ) {
        num_uniq_digits <- length(unique(strsplit(candidate, split=NULL)[[1]]))
        if ( num_uniq_digits == nchar(candidate) ) {
            filtered <- c(filtered, candidate)
        }
    }

    return(filtered)
}

pandigitals <- vector()
multiples <- get_divisible_by(2)
for ( first in seq.int(0, 9) ) {
    pandigitals <- c(pandigitals, paste0(as.character(first), multiples))
}
pandigitals <- filter_not_pandigital(pandigitals)
for ( prime in c(3, 5, 7, 11, 13, 17) ) {
    multiples <- get_divisible_by(prime)
    prefixes <- substr(multiples, 1, 2)
    next_char <- substr(multiples, 3, 3)
    new_pandigitals <- vector()
    for ( pandigital in pandigitals ) {
        suffix <- substr(pandigital, nchar(pandigital) - 1, nchar(pandigital))
        next_chars <- next_char[prefixes == suffix]
        if ( length(next_chars) ) {
            new_pandigitals <- c(new_pandigitals,
                                 paste0(pandigital, next_chars))
        }
    }
    pandigitals <- filter_not_pandigital(new_pandigitals)
}

cat(sum(as.numeric(pandigitals)), fill=TRUE)

File: 43.js
#!/usr/bin/env node

// Sub-string divisibility

var pandigital = [];
var multiples = get_divisible_by(2);
for (var first = 0; first < 10; first++) {
    for (var i = 0; i < multiples.length; i++) {
        pandigital.push(first.toString() + multiples[i]);
    }
}
pandigital = filter_not_pandigital(pandigital);
var primes = [3, 5, 7, 11, 13, 17];
for (var i = 0; i < primes.length; i++) {
    multiples = get_divisible_by(primes[i]);
    var next_char_for = {};
    for (var j = 0; j < multiples.length; j++) {
        var prefix = multiples[j].substr(0, 2);
        if (!next_char_for[prefix]) {
            next_char_for[prefix] = [];
        }
        next_char_for[prefix].push(multiples[j].slice(-1));
    }
    var new_pandigital = [];
    for (var j = 0; j < pandigital.length; j++) {
        var suffix = pandigital[j].slice(-2);
        if (next_char_for[suffix]) {
            for (var k = 0; k < next_char_for[suffix].length; k++) {
                new_pandigital.push(pandigital[j] + next_char_for[suffix][k]);
            }
        }
    }
    pandigital = filter_not_pandigital(new_pandigital);
}

var sum = 0;
for (var i = 0; i < pandigital.length; i++) {
    sum += parseInt(pandigital[i]);
}

console.log(sum);

// Get 3 digit numbers divisible by a specified number
function get_divisible_by(divisor) {
    var numbers = [];
    var number = 0;
    while (number + divisor < 1000) {
        number += divisor;
        str_number = ('000' + number).slice(-3);
        var digits = unique(str_number.split(''));
        if (digits.length != str_number.length) {
            continue;
        }
        numbers.push(str_number);
    }

    return numbers;
}

// Filter out non-pandigital numbers
function filter_not_pandigital(candidates) {
    filtered = [];
    for (var i = 0; i < candidates.length; i++) {
        var digits = unique(candidates[i].split(''));
        if (digits.length != candidates[i].length) {
            continue;
        }
        filtered.push(candidates[i]);
    }

    return filtered;
}

function unique(array) {
    return array.reduce(function(prev, cur) {
        if (prev.indexOf(cur) < 0) {
            prev.push(cur);
        }
        return prev;
    }, []);
}
